apps/site/lib/site_web/views/bus_stop_change_view.ex

```diff
@@ -1,80 +1,70 @@
 defmodule SiteWeb.BusStopChangeView do
   use SiteWeb, :view
-  alias Alerts.Alert
-  alias Routes.Route
+  alias SiteWeb.AmbiguousAlert
   alias Stops.Stop
 
-  @type alerts_by_stop :: [{Stop.t(), [%Alert{}]}]
+  @type keyname :: String.t()
+  @type alerts_with_keys :: [{keyname, AmbiguiousAlert.t()}]
+  @type alerts_by_key :: [{keyname, [AmbiguiousAlert.t()]}]
 
-  @spec sorted_by_start_date([%Alert{}]) :: [%Alert{}]
-  def sorted_by_start_date(alerts) do
-    Enum.sort_by(
-      alerts,
-      fn %Alert{active_period: [{start_date, _} | _]} ->
-        start_date
-      end
-    )
-  end
+  defdelegate affected_routes(alert), to: AmbiguousAlert
+  defdelegate related_stops(alert), to: AmbiguousAlert
+  defdelegate alert_start_date(alert), to: AmbiguousAlert
+  defdelegate alert_municipality(alert), to: AmbiguousAlert
+  defdelegate time_range(alert), to: AmbiguousAlert
+  defdelegate alert_item(alert, conn), to: AmbiguousAlert
 
-  @spec grouped_by_municipality([%Alert{}]) :: %{String.t() => alerts_by_stop}
-  def grouped_by_municipality(alerts) do
-    grouped_by_stop(alerts)
-    |> Enum.group_by(fn
-      {%Stop{municipality: municipality}, _alerts} -> municipality
-      {_stopname, _alerts} -> nil
-    end)
+  def keyed_by_municipality(alerts) do
+    alerts
+    |> Enum.map(&{alert_municipality(&1), &1})
+    |> sort_and_adjust_values(&keyed_by_stop/1)
   end
 
-  @spec grouped_by_stop([%Alert{}]) :: alerts_by_stop
-  def grouped_by_stop(alerts) do
-    alerts
-    |> Enum.flat_map(fn alert ->
-      alert
-      |> related_stops()
-      |> Enum.map(&{&1, alert})
-    end)
-    |> Enum.group_by(&elem(&1, 0), &elem(&1, 1))
-    |> Enum.sort_by(fn
-      {%Stop{name: name}, _alerts} -> name
-      {stopname, _alerts} -> stopname
-    end)
-    |> Enum.map(fn {stop, alerts} ->
-      {stop, sorted_by_start_date(alerts)}
+  @spec keyed_by_stop(alerts_with_keys) :: alerts_by_key
+  defp keyed_by_stop(keyed_alerts) do
+    keyed_alerts
+    |> Enum.flat_map(&alert_with_related_stops/1)
+    |> sort_and_adjust_values(fn alerts_list ->
+      Enum.sort_by(alerts_list, &alert_start_date/1)
```

Comment:
suggestion: when you're sorting by structs, you need to provide the name of the struct to ensure that they're not compared as maps.

```suggestion
      Enum.sort_by(alerts_list, &alert_start_date/1, DateTime)
```

There's more about why you need this in the `Enum.sort_by/3` [documentation](https://hexdocs.pm/elixir/Enum.html#sort_by/2).